src/catalog/src/persist.rs

```diff
@@ -0,0 +1,1079 @@
+        proto::ServerConfigurationValue,
+    ),
+    SystemObjectMapping(proto::GidMappingKey, proto::GidMappingValue),
+    SystemPrivilege(proto::SystemPrivilegesKey, proto::SystemPrivilegesValue),
+    Timestamp(proto::TimestampKey, proto::TimestampValue),
+}
+
+// TODO(jkosh44) As an initial implementation we simply serialize and deserialize everything as a
+// JSON. Before enabling in production, we'd like to utilize Protobuf to serialize and deserialize
+// everything.
+impl Codec for StateUpdateKind {
+    type Schema = VecU8Schema;
+
+    fn codec_name() -> String {
+        "StateUpdateProto".to_string()
+    }
+
+    fn encode<B>(&self, buf: &mut B)
+    where
+        B: bytes::BufMut,
+    {
+        let bytes = serde_json::to_vec(&self).expect("failed to encode StateUpdate");
+        buf.put(bytes.as_slice());
+    }
+
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
+        serde_json::from_slice(buf).map_err(|err| err.to_string())
+    }
+}
+
+impl mz_persist_types::columnar::Schema<StateUpdateKind> for VecU8Schema {
+    type Encoder<'a> = SimpleEncoder<'a, StateUpdateKind, Vec<u8>>;
+
+    type Decoder<'a> = SimpleDecoder<'a, StateUpdateKind, Vec<u8>>;
+
+    fn columns(&self) -> DynStructCfg {
+        SimpleSchema::<StateUpdateKind, Vec<u8>>::columns(&())
+    }
+
+    fn decoder<'a>(&self, cols: ColumnsRef<'a>) -> Result<Self::Decoder<'a>, String> {
+        SimpleSchema::<StateUpdateKind, Vec<u8>>::decoder(cols, |val, ret| {
+            *ret = StateUpdateKind::decode(val).expect("should be valid StateUpdateKind")
+        })
+    }
+
+    fn encoder<'a>(&self, cols: ColumnsMut<'a>) -> Result<Self::Encoder<'a>, String> {
+        SimpleSchema::<StateUpdateKind, Vec<u8>>::push_encoder(cols, |col, val| {
+            let mut buf = Vec::new();
+            StateUpdateKind::encode(val, &mut buf);
+            mz_persist_types::columnar::ColumnPush::<Vec<u8>>::push(col, &buf)
+        })
+    }
+}
+
+/// Handles and metadata needed to interact with persist.
+///
+/// Production users should call [`Self::_expire`] before dropping a [`PersistHandle`] so that it
+/// can expire its leases. If/when rust gets AsyncDrop, this will be done automatically.
+#[derive(Debug)]
+pub struct PersistHandle {
+    /// Write handle to persist.
+    write_handle: WriteHandle<StateUpdateKind, (), Timestamp, Diff>,
+    /// Read handle to persist.
+    read_handle: ReadHandle<StateUpdateKind, (), Timestamp, Diff>,
+}
+
+impl PersistHandle {
+    /// Create a new [`PersistHandle`] to the catalog state associated with `environment_id`.
+    pub(crate) async fn new(persist_client: PersistClient, environment_id: Uuid) -> PersistHandle {
+        // TODO(jkosh44) Using the environment ID directly is sufficient for correctness purposes.
+        // However, for observability reasons, it would be much more readble if the shard ID was
+        // constructed as `format!("s{}CATALOG", hash(environment_id))`.
```

Comment:
not sure if this is possible, but the `shard_name` stuff below should help! that name shows up in the persist metrics and grafana dash